mdbook/src/chapter_4/chapter_4_2.md

```diff
@@ -46,9 +47,10 @@ As a computation proceeds, operators may perform three classes of action:
 
 The results of these actions are a stream of changes to the occurrences of capabilities at each location in the dataflow graph. As input messages are consumed, capabilies located at the corresponding `Target` input port are decremented, and capabilities at the operators `Source` output ports are incremented. Cloning, downgrading, and dropping capabilities changes the counts at each of the operators corresponding `Source` output ports. Sending messages results in increments to the counts at the `Target` ports of each `Target` connected to the `Source` from which the message is sent.
 
-Concretely, a batch of changes has the form `(Vec<(Source, Time, i64)>, Vec<(Target, Time, i64)>), indicating the increments and decrements for each time, at each dataflow location. Importantly we must keep these batches intact; the safety of the progress tracking protocol relies on not communicating half-formed progress messages (for example, consuming an input message but forgetting to indicate the acquisition of its capability).
+Concretely, a batch of changes has the form `(Vec<(Source, Time, i64)>, Vec<(Target, Time, i64)>), indicating the increments and decrements for each time, at each dataflow location. Importantly we must keep these batches intact; the safety of the progress tracking protocol relies on not communicating half-formed progress messages (for example, consuming an input message but forgetting to indicate the acquisition of its capability).[Do progress change batches
+always correspond to information corresponding to some union of "data batches"?]
 
-Each worker reliably broadcasts the stream of progress change batches to all workers in the system (including itself) along FIFO channels. At any point in time, each worker has seen an arbitrary prefix of the sequence of progress change batches produced by each other worker, and it is expected that as time proceeds each worker eventually sees every prefix of each sequence.
+Each worker reliably broadcasts the stream of progress change batches to all workers in the system (including itself) along FIFO channels. At any point in time, each worker has seen an arbitrary prefix of the sequence of progress change batches produced by each other worker, and it is expected that as time proceeds each worker eventually sees every prefix of each sequence[What ordering guarantees do you have here? This looks like the kind of thing that's annoying to get completely right in practice].
```

Comment:
The ordering guarantees should be just that each point-to-point (worker to worker) progress update channel is FIFO and at most once. If you make it "reliable" then you get progress too. There is no requirement that these messages be correctly interleaved with data or anything like that. In fact, if you do some silly looking changes to the progress protocol, you can drop the FIFO requirement (if you change it to disallow cloning of capabilities, and require consuming capabilities when sending a message; this would mean building in a magical 100000x inflation on received capabilities, or something similar, but it would remove the requirement that progress update batches be FIFO (and later in the text "prefix of progress update batches" could be replaced with "subset of progress update batches" ).